lib/fluent/plugin_helper/child_process.rb

```diff
@@ -0,0 +1,206 @@
+#
+# Fluentd
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'fluent/plugin_helper/thread'
+require 'fluent/plugin_helper/timer'
+
+module Fluent
+  module PluginHelper
+    module ChildProcess
+      include Fluent::PluginHelper::Thread
+      include Fluent::PluginHelper::Timer
+
+      CHILD_PROCESS_LOOP_CHECK_INTERVAL = 0.2 # sec
+      CHILD_PROCESS_DEFAULT_KILL_TIMEOUT = 60 # sec
+
+      # stop     : [-]
+      # shutdown : send TERM to all child processes
+      # close    : close all I/O objects for child processes, send TERM and then KILL after timeout
+      # terminate: [-]
+
+      attr_reader :_child_process_processes # for tests
+
+      def child_process_execute(
+          title, command,
+          arguments: nil, subprocess_name: nil, interval: nil, immediate: false, parallel: false,
+          read: true, write: true, internal_encoding: 'utf-8', external_encoding: 'utf-8', &block
+      )
+        raise ArgumentError, "BUG: title must be a symbol" unless title.is_a? Symbol
```

Comment:
There are lots of `raise ArgumentError, "BUG: message" if cond`.
How about providing check helper in Plugin like Guava's preconditions?

```rb
Plugin.check_param(cond, 'message');
# or 
PluginHelper.check_param(cond, 'message');
```